False Positive "G602: slice index out of range" over "range" of array with given size

Since my last golangci-lint update - using the latest gosec version 2.23.0 -, a false positive G602 occurs when `range` over the identifier of an array with a fixed size.

```
$ go version -m ~/go/bin/golangci-lint | grep gosec
        dep     github.com/securego/gosec/v2    v2.23.0 h1:h4TtF64qFzvnkqvsHC/knT7YC5fqyOCItlVR8+ptEBo=
```

Take the following code as a minimal example.

```go
package main

import "fmt"

func main() {
	var arr [8]int

	for i := range arr {
		arr[i] = i
		fmt.Printf("i=%d\tlen(arr)=%d\tcap(arr)=%d\tarr[%d]=%d\n", i, len(arr), cap(arr), i, arr[i])
	}
}

```

```
$ go run main.go
i=0     len(arr)=8      cap(arr)=8      arr[0]=0
i=1     len(arr)=8      cap(arr)=8      arr[1]=1
i=2     len(arr)=8      cap(arr)=8      arr[2]=2
i=3     len(arr)=8      cap(arr)=8      arr[3]=3
i=4     len(arr)=8      cap(arr)=8      arr[4]=4
i=5     len(arr)=8      cap(arr)=8      arr[5]=5
i=6     len(arr)=8      cap(arr)=8      arr[6]=6
i=7     len(arr)=8      cap(arr)=8      arr[7]=7

$ ~/go/bin/golangci-lint run -E gosec
main.go:9:6: G602: slice index out of range (gosec)
                arr[i] = i
                   ^
main.go:10:91: G602: slice index out of range (gosec)
                fmt.Printf("i=%d\tlen(arr)=%d\tcap(arr)=%d\tarr[%d]=%d\n", i, len(arr), cap(arr), i, arr[i])
                                                                                                        ^
2 issues:
* gosec: 2
```

Please pardon me if I am wrong, but in my understanding this `range arr` call returns all indices. These indices are then used to access the array - in the example ranging from 0 to 7. I cannot see an index out of range.

Btw, when I change the `range arr` to one of the following, the program behaves identical and gosec does not complain any longer:
- `for i := range len(arr) { ... }`
- `for i := range arr[:] { ... }`

Thanks a lot!

--- a/analyzers/slice_bounds.go
+++ b/analyzers/slice_bounds.go
@@ -576,6 +576,37 @@ func (s *sliceBoundsState) extractIntValueIndexAddr(refinstr *ssa.IndexAddr, sli
 	base, offset := decomposeIndex(refinstr.Index)
 	var sliceIncr int
 
+	canNormalizeToBase := func(bin *ssa.BinOp) bool {
+		if bin == nil || refinstr == nil {
+			return false
+		}
+		binBlock := bin.Block()
+		idxBlock := refinstr.Block()
+		if binBlock == nil || idxBlock == nil {
+			return false
+		}
+		if binBlock != idxBlock {
+			return true
+		}
+		binPos := -1
+		idxPos := -1
+		for i, ins := range binBlock.Instrs {
+			if ins == bin {
+				binPos = i
+			}
+			if ins == refinstr {
+				idxPos = i
+			}
+			if binPos >= 0 && idxPos >= 0 {
+				break
+			}
+		}
+		if binPos < 0 || idxPos < 0 {
+			return false
+		}
+		return binPos < idxPos
+	}
+
 	// Case 1: Base is a constant (e.g., s[0+3])
 	if val, ok := GetConstantInt64(base); ok {
 		finalIdx := int(val) + offset
@@ -640,16 +671,15 @@ func (s *sliceBoundsState) extractIntValueIndexAddr(refinstr *ssa.IndexAddr, sli
 							}
 							maxV := limit + incr
 
-							// If the limit is on 'next' (i+1 < limit), it still bounds 'i'
-							// In 'range n', i reaches n-1.
-							// Here we use a heuristic: if we find an upper bound, check it.
+							// If the limit is found on an incremented value (next or nBase != p),
+							// normalize it back to the base loop variable before applying index offset.
+							boundAdjust := 0
+							if (v == next && base != next && canNormalizeToBase(bin)) || (v == nBase && nBase != p && base != nBase) {
+								boundAdjust = -nOffset
+							}
+
 							if bound == lowerUnbounded || bound == upperBounded {
-								// Correct the max value of 'base' based on where the limit was found
-								finalMaxV := maxV
-								if v == nBase && nBase != p {
-									// if i + nOffset < limit, then i < limit - nOffset
-									finalMaxV = maxV - nOffset
-								}
+								finalMaxV := maxV + boundAdjust
 								if !isSliceIndexInsideBounds(sliceCap+sliceIncr, finalMaxV+offset) {
 									return finalMaxV + offset, nil
 								}
@@ -661,10 +691,11 @@ func (s *sliceBoundsState) extractIntValueIndexAddr(refinstr *ssa.IndexAddr, sli
 							incr := -1 // extractLenBound only handles LSS for now
 							maxV := limit + off + incr
 
-							finalMaxV := maxV
-							if v == nBase && nBase != p {
-								finalMaxV = maxV - nOffset
+							boundAdjust := 0
+							if (v == next && base != next && canNormalizeToBase(bin)) || (v == nBase && nBase != p && base != nBase) {
+								boundAdjust = -nOffset
 							}
+							finalMaxV := maxV + boundAdjust
 							if !isSliceIndexInsideBounds(sliceCap+sliceIncr, finalMaxV+offset) {
 								return finalMaxV + offset, nil
 							}

--- a/testutils/g602_samples.go
+++ b/testutils/g602_samples.go
@@ -696,4 +696,23 @@ func main() {
 	_ = arr[4]
 }
 `}, 0, gosec.NewConfig()},
+	// Issue #1525: G602 false positive for array index in range-over-array loops
+	{[]string{`
+package main
+func main() {
+	var arr [8]int
+	for i := range arr {
+		arr[i] = i
+	}
+}
+`}, 0, gosec.NewConfig()},
+	{[]string{`
+package main
+func main() {
+	var arr [8]int
+	for i := range arr {
+		_ = arr[i+1]
+	}
+}
+`}, 1, gosec.NewConfig()},
 }

